pkg/cloudprovider/servicecontroller/servicecontroller.go

```diff
@@ -0,0 +1,426 @@
+	// TODO: Add a compressor that intelligently squashes together updates.
+	keyLister := cache.KeyListerFunc(func() []string { return s.listKeys() })
+	serviceQueue := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, keyLister)
+	lw := cache.NewListWatchFromClient(s.kubeClient.(*client.Client), "services", api.NamespaceAll, fields.Everything())
+	cache.NewReflector(lw, &api.Service{}, serviceQueue, 0).Run()
+	// TODO: Add proper retries rather than just re-adding to the queue?
+	for {
+		newItem := serviceQueue.Pop()
+		deltas, ok := newItem.(cache.Deltas)
+		if !ok {
+			glog.Errorf("Received object from service watcher that wasn't Deltas: %+v", newItem)
+		}
+		delta := deltas.Newest()
+		if delta == nil {
+			glog.Errorf("Received nil delta from watcher queue.")
+			continue
+		}
+		err := s.processDelta(delta)
+		if err != nil {
+			// Add the failed service back to the queue so we'll retry it.
+			glog.Error(err)
+			time.Sleep(5 * time.Second)
+			serviceQueue.AddIfNotPresent(deltas)
+			break
+		}
+	}
+}
+
+// Returns an error if processing the delta should be retried, otherwise returns nil.
+func (s *ServiceController) processDelta(delta *cache.Delta) error {
+	service, ok := delta.Object.(*api.Service)
+	if !ok {
+		// If the DeltaFIFO saw a key in our cache that it didn't know about, it
+		// can send a deletion with an unknown state. Grab the service from our
+		// cache for deleting.
+		key, ok := delta.Object.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			glog.Errorf("Delta contained object that wasn't a service or a deleted key: %+v", delta)
+			return nil
+		}
+		service, ok = s.getService(key.Key)
+		if !ok {
+			glog.Errorf("Service %s not in cache even though the watcher thought it was. Ignoring the deletion.", key)
+			return nil
+		}
+		delta.Object = service
+	}
+	glog.V(2).Infof("Got new %s delta for service: %+v", delta.Type, service)
+
+	// TODO: Make this more parallel. The only things that need to serialized
+	// are changes to services with the same namespace and name.
+	// TODO: Handle added, updated, and sync differently?
+	switch delta.Type {
+	case cache.Added:
+		fallthrough
+	case cache.Updated:
+		fallthrough
+	case cache.Sync:
+		return s.createLoadBalancerIfNeeded(service)
+	case cache.Deleted:
+		return s.handleDelete(service)
+	default:
+		glog.Errorf("Unexpected delta type: %v", delta.Type)
+	}
+	return nil
+}
+
+func (s *ServiceController) createLoadBalancerIfNeeded(service *api.Service) error {
+	namespacedName, err := cache.MetaNamespaceKeyFunc(service)
+	if err != nil {
+		glog.Errorf("Couldn't generate namespaced name for service: %v", err)
+		return nil
+	}
+
+	cachedService, cached := s.getService(namespacedName)
+	if cached && !needsUpdate(cachedService, service) {
+		glog.Infof("LB already exists and doesn't need update for service %s", namespacedName)
+		return nil
+	}
+	if cached {
+		// If the service already exists but needs to be updated, delete it so that
+		// we can recreate it cleanly.
```

Comment:
Won't this create an unnecessarily long break in network connectivity to the service?  Can it not rather be updated in place?